core/src/main/scala/kafka/log/Log.scala

```diff
@@ -485,7 +485,13 @@ class Log(val dir: File,
     // but if that segment doesn't contain any messages with an offset greater than that
     // continue to read from successive segments until we get some messages or we reach the end of the log
     while(entry != null) {
-      val fetchInfo = entry.getValue.read(startOffset, maxOffset, maxLength)
+      val maxPosition = {
+        if (entry == segments.lastEntry)
+          Some(nextOffsetMetadata.relativePositionInSegment.toLong)
```

Comment:
There is a potential race condition. It's possible that the last segment changes after the test in line 489. Then the maxPosition could be on a newly created segment. This may lead to a negative length in line 139 in LogSegment. Currently, FileMessageSet.read() will throw an exception on negative length.